### PR TITLE
Allow Upper case and _ as first character of odin attributes, as per …

### PR DESCRIPTION
**Pull request by pieterbos**
…older spec

### CODE REVIEW

File: grammars/src/main/antlr/BaseLexer.g4 [hunk 1]
```diff
@@ -151,8 +151,9 @@ fragment LABEL : ALPHA_CHAR (NAME_CHAR|URI_PCT_ENCODED)* ;
 
 GUID : HEX_DIGIT+ '-' HEX_DIGIT+ '-' HEX_DIGIT+ '-' HEX_DIGIT+ '-' HEX_DIGIT+ ;
 
-ALPHA_UC_ID : ALPHA_UCHAR WORD_CHAR* ;           // used for type ids
-ALPHA_LC_ID : ALPHA_LCHAR WORD_CHAR* ;           // used for attribute / method ids
+ALPHA_UC_ID :   ALPHA_UCHAR WORD_CHAR* ;   // used for type ids
+ALPHA_LC_ID :   ALPHA_LCHAR WORD_CHAR* ;   // used for attribute / method ids
+ALPHA_UNDERSCORE_ID : '_' WORD_CHAR* ;     // usually used for meta-model ids
```
Comment (J3173) [inline grammars/src/main/antlr/BaseLexer.g4:156]:
```suggestion
UNDERSCORE_ID : '_' WORD_CHAR* ;     // usually used for meta-model ids
```
As the format seems to be `firstchar_ID`.

**Review by J3173 — CHANGES_REQUESTED** [review]
You say "as per older spec". Do we need these changes in the ADL2 grammar then or should only changing the 1.4 grammar be sufficient?

**Comment by pieterbos** [comment]
This needs to be done both in ADL 1.4 and ADL 2 - otherwise these ADL 1.4 archetypes cannot be converted to ADL 2. I couldn't find why this was changed, but introducing a backwards incompatible change in an object serialization language such as ODIN is just a bad idea, regardless of ADL version.

File: grammars/src/main/antlr/BaseLexer.g4 (hunk 1 above)
Comment (pieterbos) [reply grammars/src/main/antlr/BaseLexer.g4:156]: I copied this from changes in the 'official' grammar at https://github.com/openEHR/adl-antlr/blob/master/src/main/antlr/adl2/base_lexer.g4#L150 , and I would like to keep this aligned as much as possible.

Comment (wolandscat) [reply grammars/src/main/antlr/BaseLexer.g4:156]: This is just changing the lexer rule id? Not sure I see the value - LC_ID = 'lower case id', UC_ID = 'upper case id'; 'UNDERSCORE_ID' = underscore id. If we want nicer names I'd rather do that in the major revamp I am working on.

Comment (J3173) [reply grammars/src/main/antlr/BaseLexer.g4:156]: > I copied this from changes in the 'official' grammar at https://github.com/openEHR/adl-antlr/blob/master/src/main/antlr/adl2/base_lexer.g4#L150 , and I would like to keep this aligned as much as possible.

Agreed.

> This is just changing the lexer rule id? Not sure I see the value - LC_ID = 'lower case id', UC_ID = 'upper case id'; 'UNDERSCORE_ID' = underscore id. If we want nicer names I'd rather do that in the major revamp I am working on.

I interpreted the rule names as: "ALPHA_UC_ID": ID starting with an ALPHA_UC (upper case letter), "ALPHA_LC_ID": ID starting with an ALPHA_LC (lower case letter).

Then "ALPHA_UNDERSCORE_ID" would be ID starting with an ALPHA_UNDERSCORE, but that is strange. For me, "UNDERSCORE_ID" makes more sense: ID starting with an UNDERSCORE.

But if it is already in use, maybe changing it isn't worth it.

**Review by J3173 — APPROVED** [review]

**Pull request merged**